### PR TITLE
Take observation time from GTI table

### PR DESCRIPTION
**Pull request by dcfidalgo**
This PR addresses issues #1866 and #1504.

The observation time will now be calculated using the GTI table pointed to in the HDU index file.
For backward compatibility, it will take the observation time from the Obs index file if there is no GTI table available.

Maybe we could add a warning that taking the obs time from the Obs index file is deprecated and will raise an error in the future?

I checked the data sets we have in gammapy-extra for differences between the live-times (computed from the GTI table and taken from the obs index file). Only the data sets
* `datasets/hess-crab4-hd-hap-prod2` and
* `test_datasets/cube/data`

show a deviation of about 0.1-0.5%. For the rest of the data sets there is basically no difference.
For `datasets/hess-crab4-pa` there is no GTI table available.

### CODE REVIEW

**Comment by dcfidalgo** [comment]
We just discussed this PR in the call. Conclusion: rather than changing the `observation_time_duration` and `observation_live_time_duration` property, it would be better to modify the `gti` property in a way that it creates a GTI table on the fly if necessary.

I will make this change here in this PR.

**Comment by dcfidalgo** [comment]
I introduced a `_create_missing_gti()` method, that is called when no GTI table is present in the HDU index file. Also added a test for this method.

Also, `observation_time_duration` and `observation_live_time_duration` are no "lazyproperties" anymore, since when applying a time filter, these values can change. So i think it is safer to not cache here.

@cdeil From my side this would be ready to merge.

**Comment by dcfidalgo** [comment]
Just did a rebase on the master, after Axels message on slack.

File: gammapy/data/observations.py [hunk 1]
```diff
@@ -356,6 +361,32 @@ def check(self, checks="all"):
...
+
+        This method can be removed once GTI tables are explicitly required in Gammapy.
+        """
+        header = OrderedDict(
+            EXTNAME="GTI",
+            HDUCLASS="GADF",
+            HDUDOC="https://github.com/open-gamma-ray-astro/gamma-astro-data-formats",
+            HDUVERS="0.2",
+            HDUCLAS1="GTI",
+            MJDREFI=self.data_store.obs_table.meta["MJDREFI"],
```
Comment (cdeil) [inline gammapy/data/observations.py]: We already use `time_ref_from_dict(self.data_store.obs_table.meta)` in other places in this file. Suggest to use it here also to get consistent behaviour.

File: gammapy/data/tests/test_observations.py [hunk 2]
```diff
@@ -38,6 +38,24 @@ def test_data_store_observation(data_store):
...
 
+@requires_data("gammapy-extra")
+def test_create_missing_gti():
+    """Test the DataStoreObservation._create_missing_gti() method.
+
+    For the 'hess-dl3-dr1' data set, the GTI tables contain only the 'TSTART' and 'TSTOP' from the 'event list' and
+    'obs index' header. The `DataStoreObservation._create_missing_gti()` method creates a GTI table on-the-fly using
+    exactly those values.
+    """
+    ds = data_store()
```
Comment (cdeil) [inline gammapy/data/tests/test_observations.py]: `data_store` is a session-scoped fixture. Fixtures shouldn't be called, they should be injected into tests. (see other existing uses of this fixture). This will give a warning or error with the latest pytest release.

Suggest to put the `_create_missing_gti` method as `ObservationTable.make_gti_table(obs_id)` and add a test for it there without doing something complex like modifying a DataStore fixture object. Given that this is just a small one-line workaround I think it's fine to have no extra unit test from the `DataStoreObservation` side.

File: gammapy/data/tests/test_observations.py [hunk 3]
```diff
@@ -38,6 +38,24 @@ def test_data_store_observation(data_store):
...
+    obs_time_1 = ds.obs(20136).observation_time_duration
+
+    ds = data_store()  # DataStore saves some lazyproperties when accessing a hdu location -> we create a new one
+    ds.hdu_table.remove_row(1)  # Remove reference to GTI table -> GTI table will be created on-the-fly
+    obs_time_2 = ds.obs(20136).observation_time_duration
+
+    assert obs_time_1 == obs_time_2
```
Comment (cdeil) [inline gammapy/data/tests/test_observations.py]: Exact asserts on floats probably isn't a good idea.

Do an allclose assert for floats, and use one of our helper functions if it makes the assert allclose simpler:
https://docs.gammapy.org/0.8/utils/index.html#id11

**Comment by dcfidalgo** [comment]
@cdeil Thanks for the comments!

I moved the method to the `ObservationTable` class. For the test i had to modify a bit the test observation table to reflect the [time header format](https://gamma-astro-data-formats.readthedocs.io/en/latest/general/time.html#time-formats).

I left a small test in the `test_observations.py` that just checks if we correctly catch the case when a GTI table is missing in the HDU Index file.

>Given that here all the remaining lazy properties (except possibly GTI) just access obs_info, which is just a row from a table, i.e. very fast, doesn't access the disk, you could either remove all lazyproperty completely in this file, or only keep them for the obs_info and gti properties. Or just leave as-is for now, as something to be figured out in a follow-up PR.

For now i simply removed all lazy properties, maybe we can discuss a possible optimization in a follow-up PR.

File: gammapy/data/tests/test_observations.py (hunk 3 above)
Comment (dcfidalgo) [reply gammapy/data/tests/test_observations.py]: Done! See the new test in `test_obs_table.py`

File: gammapy/data/observations.py (hunk 1 above)
Comment (dcfidalgo) [reply gammapy/data/observations.py]: `time_ref_from_dict` already returns a `Time` object, whereas here i rather need the int and float values.

File: gammapy/data/obs_table.py [hunk 4]
```diff
@@ -351,6 +352,41 @@ def select_observations(self, selection=None):
...
+            HDUCLAS1="GTI",
+            MJDREFI=self.meta["MJDREFI"],
+            MJDREFF=self.meta["MJDREFF"],
+            TIMEUNIT=self.meta.get("TIMEUNIT", "s"),
+            TIMESYS=self.meta["TIMESYS"],
+            TIMEREF=self.meta.get("TIMEREF", "LOCAL"),
+        )
+
+        idx = self.get_obs_idx(obs_id)[0]
+        start = self["TSTART"][idx].astype("float64")
```
Comment (cdeil) [inline gammapy/data/obs_table.py]: I don't think an `astype("float64")` here can be helpful; the spec says it should be 64-bit, and if it isn't going 32 -> 64 at this point doesn't help, if the precision is already low, the result will be the same.

I'll do a small edit here to ensure the unit in the GTI table is "s". It should already be in the obs table, but given that you hardcode `self.meta.get("TIMEUNIT", "s")` above, I think it's a little less error-prone to also go to "s" for the value here.

File: gammapy/data/tests/test_observations.py [hunk 5]
```diff
@@ -38,6 +34,26 @@ def test_data_store_observation(data_store):
     assert_skycoord_allclose(obs.target_radec, c)
 
 
+@requires_data("gammapy-extra")
+def test_missing_gti(data_store):
+    """This tests the case when a GTI table is missing in the HDU index file.
+
+    For backward compatibility we create a GTI table on-the-fly by means of the Obs Index file
+    TODO: This method can be removed once GTI tables are explicitly required in Gammapy.
```
Comment (cdeil) [inline gammapy/data/tests/test_observations.py]: @dcfidalgo - I'll remove this test already now. I've had very bad experiences with tests that modify fixture objects, and then depending on test execution order other tests broke and it was very hard to debug. I also don't like to have such complex tests where state is backed up and restored.
Overall this is an edge case that should go away soon, I think it's OK to leave this one line untested.
